substrate/client/network/src/litep2p/mod.rs

```diff
@@ -0,0 +1,872 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! `NetworkBackend` implementation for `litep2p`.
+
+use crate::{
+	config::{
+		FullNetworkConfiguration, IncomingRequest, NodeKeyConfig, NotificationHandshake, Params,
+		SetConfig, TransportConfig,
+	},
+	error::Error,
+	event::{DhtEvent, Event},
+	litep2p::{
+		discovery::{Discovery, DiscoveryEvent},
+		peerstore::Peerstore,
+		service::{Litep2pNetworkService, NetworkServiceCommand},
+		shim::{
+			bitswap::BitswapServer,
+			notification::{
+				config::{NotificationProtocolConfig, ProtocolControlHandle},
+				peerset::PeersetCommand,
+			},
+			request_response::{RequestResponseConfig, RequestResponseProtocol},
+		},
+	},
+	multiaddr::{Multiaddr, Protocol},
+	peer_store::PeerStoreProvider,
+	protocol,
+	service::{
+		metrics::{register_without_sources, MetricSources, Metrics, NotificationMetrics},
+		out_events,
+		traits::{BandwidthSink, NetworkBackend, NetworkService},
+	},
+	NetworkStatus, NotificationService, ProtocolName,
+};
+
+use codec::Encode;
+use futures::StreamExt;
+use libp2p::kad::RecordKey;
+use litep2p::{
+	config::ConfigBuilder,
+	crypto::ed25519::{Keypair, SecretKey},
+	executor::Executor,
+	protocol::{
+		libp2p::{bitswap::Config as BitswapConfig, kademlia::QueryId},
+		request_response::ConfigBuilder as RequestResponseConfigBuilder,
+	},
+	transport::{
+		tcp::config::Config as TcpTransportConfig,
+		websocket::config::Config as WebSocketTransportConfig,
+	},
+	Error as Litep2pError, Litep2p, Litep2pEvent, ProtocolName as Litep2pProtocolName,
+};
+use parking_lot::RwLock;
+use prometheus_endpoint::Registry;
+
+use sc_client_api::BlockBackend;
+use sc_network_common::{role::Roles, ExHashT};
+use sc_network_types::PeerId;
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver};
+use sp_runtime::traits::Block as BlockT;
+
+use std::{
+	cmp,
+	collections::{HashMap, HashSet},
+	fs,
+	future::Future,
+	io, iter,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicUsize, Ordering},
+		Arc,
+	},
+	time::Duration,
+};
+
+mod discovery;
+mod peerstore;
+mod service;
+mod shim;
+
+/// Litep2p bandwidth sink.
+struct Litep2pBandwidthSink {
+	sink: litep2p::BandwidthSink,
+}
+
+impl BandwidthSink for Litep2pBandwidthSink {
+	fn total_inbound(&self) -> u64 {
+		self.sink.inbound() as u64
+	}
+
+	fn total_outbound(&self) -> u64 {
+		self.sink.outbound() as u64
+	}
+}
+
+/// Litep2p task executor.
+struct Litep2pExecutor {
+	/// Executor.
+	executor: Box<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>) + Send + Sync>,
+}
+
+impl Executor for Litep2pExecutor {
+	fn run(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+		(self.executor)(future)
+	}
+
+	fn run_with_name(&self, _: &'static str, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+		(self.executor)(future)
+	}
```

Comment:
Is this planned to be implemented?

Reply:
No plan, I was using this during debugging but we don't use it in Polkadot.